### PR TITLE
Creating a caching key for queries the H6 kind of way

### LINKED ISSUE #16332: Hibernate 6 upgrade: unable to extract query parameter name in QueryCacheKey
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

storage

### Describe the bug

After upgrading to hibernate 6, the `QueryCacheKey` constructor fails due to the query parameter's name being now null. Switching to position doesn't help as that also returns a null value.

This code was put in place in the quarkus3 branch to be able to start Keycloak with the map store: (`JpaMapKeycloakTransaction.QueryCacheKey`)

```
            for (Parameter<?> parameter : query.getParameters()) {
                // HACK TO GET KC TO START WITH NEW STORE - HAS TO BE FIXED AS SOME SEARCHES USE A DIFFERENT EXPRESSIBLE IMPL
                JpaCriteriaParameter criteriaParameter = (JpaCriteriaParameter) parameter;
                SqmBasicValuedSimplePath expressible = (SqmBasicValuedSimplePath) criteriaParameter.getExpressible();
                String parameterName  = expressible.getNavigablePath().getLocalName();
                if (parameterName == null) {
                    throw new ModelException("Can't prepare query for caching as parameter doesn't have a name");
                }
                this.queryParameters.put(parameterName, ((JpaCriteriaParameter<?>) parameter).getValue());
            }
```

We need to figure out how to properly handle this as the above hack won't work in all scenarios.

### Version

quarkus3 branch

### Expected behavior

QueryCacheKey should be able to retrieve the query parameter names.

### Actual behavior

QueryCacheKey obtains names using explicit type casts that won't always work.

### How to Reproduce?

Use quarkus3 branch

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by ahus1**
Closes #16332

### CODE REVIEW

**Comment by ahus1** [comment]
@hmlnarik - this is now ready for review. A manual local test showed that this work now better when showing users related to a role. 

This only compiles and runs the unit tests, everything else is expected to fail until more of the Arquillian setup has been fixed.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:193]: Thank you @ahus1!

It seems to me that for each call of `read(QueryParameters<M> queryParameters)` there is empty map (`aliases`) created and for each root, which doesn't have an alias, it's created one. Then the alias is inserted into `aliases`.

Could you please provide an example of a query with multiple roots where alias would be found on line 185? So far I was not able to came across such a query. Thank you

**Review by hmlnarik — APPROVED** [review]

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:193]: @vramik - I'm not aware of a secondary root being used at the moment. The closes I found was in JpaRoleModelCriteriaBuilder the subquery that's being created. 

I tried to trigger this code by calling a delete on  a role, but this this failed earlier as the "->" JSON functions weren't working (that's possibly another issue that we're about to soon for the H6 upgrade).

So to summarize: At the moment it is used only for one root, and I am not sure what it would do on the subquery. If the code would not set an alias for whatever reason, it would lead to a query that can't be cached, which might turn up in a performance test eventually.  

This whole caching code is currently used to cache the realm when working via the admin UI (and REST endpoints). It is also used to cache a client if a recall correctly during refresh of a token. I verified that those caching works as expected.

It still has its justification IMHO, at the same time we might skip it if we see it adds to much complexity.

Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:193]: Makes sense, thanks. As far as I could see caching works as expected. My understating of the code above was that aliases are only stored into a map but there is never alias found in line 185. 

I just wanted to understand the purpose of the map properly so I wonder whether my understanding is correct or not.

Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:193]: @vramik - yes, your understanding is correct. As there is currently only a single root, the logic isn't really tested used yet.  Today all aliases have the "0" at the end. For that to be "1", it would need to have a root with the same name a second time. 

I was assigning a separate numbering for each root as for two roots, I wouldn't know in which order I would receive them from the set of roots. So it was better to number those wich the same name consecutively instead of numbering all aliases consecutively.

Let me know if this comment helped to get this PR towards an approval for a review, or if you'd prefer to have a short call.

Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:193]: Thanks for the comment, makes sense to me. Approving.

**Review by vramik — APPROVED** [review]

**Comment by ahus1** [comment]
@hmlnarik - thanks for looking at the model tests, I missed that those should already run with H6. The error messages were suspicious, and I updated my code. 

The way to extract the values and and to generated artificial parameter names was incomplete and fragile. I now look that the value is part of the query string. 

This now fixed the tests when I ran them locally, CI is still running. 

Please have a look. 

This is all built on the assumption that if there is no query parameter name, Hibernate will have put it into the query string - and this is a very internal Hibernate behavior. And HQL strings are not well supported in H6 anyway, as they're moving towards JPQL even internally AFAIK.

I think looking into the query string is still too fragile and error prone, I'm ready to rip out all of it. It will affect the performance of the admin UI with regards to caching of the realm entity, but with other optimizations I made it is not as important as it was earlier. If we rip it out, we'll have to wait for #13641.

Switching this to a draft state due to this.

**Comment by ahus1** [comment]
@hmlnarik / @vramik - after talking to @hmlnarik and discussing my concerns with the solution that was dependent on  internal Hibernate workings, this PR is now updated to use the `ModelCriterialBuilder` directly to prepare the query. 

As this takes out the ugly bits, this is now ready for review again. Thanks!

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java
Comment (hmlnarik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:355]: Nit: this could be checked to be `DefaultModelCriteria` already, in which case flashing could be skipped.

**Review by hmlnarik — COMMENTED** [review]
Thank you @ahus1! The changes are so good that I'd propose opening them against `main` branch instead, WDYT?

**Comment by ahus1** [comment]
@hmlnarik - I'd like to merge this to the Quarkus 3 branch first as I want to see the model tests succeed with our three PRs. 

Once it is merged to Quarkus 3, I'll prepare a another PR for main. That would then lead to some merge conflicts on rebase for the Quarkus 3 branch.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java
Comment (hmlnarik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java]: I agree

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:174]: Thank you @ahus1, I like the way it's done. Great idea.

I have just one nitpick: there is `null` assigned into `queryCacheKey` on line 173 and on 174 there is actual `QueryCacheKey`. Would it make sense:
```suggestion
        QueryCacheKey queryCacheKey = new QueryCacheKey(queryParameters, modelType);
```
?

Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:174]: Will do. As Hynek suggested, I'm now preparing a new PR against main, see #16516 (it is currently in draft)

**Pull request closed without merging**
